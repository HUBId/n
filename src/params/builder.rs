use super::types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashKind, LdeOrder, LdeParams,
    MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
use super::{ParamsError, StarkParams};

/// Builder used to assemble [`StarkParams`] with validation.
///
/// | Field | Default |
/// |-------|---------|
/// | `params_version` | `1` |
/// | `field` | [`FieldKind::Goldilocks`] |
/// | `hash` | [`HashKind::Poseidon2 { parameter_set: 0 }`] |
/// | `lde` | Blowup `8`, row-major order, coset tag `0x504152414d533031` |
/// | `fri` | `r = 4`, `queries = 30`, `domain_log2 = 22`, natural folding, `num_layers = 5` |
/// | `merkle` | Little-endian leaves, width `4`, binary arity, domain sep `0x4d4b4c5f50415231` |
/// | `transcript` | Protocol tag `0x5354524b5f50524f`, deterministic seed, challenge bounds `16..=64` |
/// | `proof` | Version `1`, `max_size_kb = 512` |
/// | `security` | Target `96` bits, slack `16` bits |
#[derive(Debug, Clone)]
pub struct StarkParamsBuilder {
    pub params_version: u16,
    pub field: FieldKind,
    pub hash: HashKind,
    pub lde: LdeParams,
    pub fri: FriParams,
    pub merkle: MerkleParams,
    pub transcript: TranscriptParams,
    pub proof: ProofParams,
    pub security: SecurityBudget,
}

impl StarkParamsBuilder {
    /// Returns a builder initialised with safe defaults.
    pub fn new() -> Self {
        Self::from_profile(BuiltinProfile::PROFILE_X8)
    }

    /// Loads one of the built-in profiles.
    ///
    /// | Profile | Description | Field | Hash | Blowup | Queries | Arity | Target Bits |
    /// |---------|-------------|-------|------|--------|---------|-------|-------------|
    /// | `PROFILE_X8` | Balanced performance profile with blowup 8 | Goldilocks | Poseidon2 (set 0) | 8 | 30 | Binary | 96 |
    /// | `PROFILE_HISEC_X16` | High security profile with blowup 16 | BN254 | Rescue (set 1) | 16 | 48 | Quaternary | 128 |
    pub fn from_profile(profile: BuiltinProfile) -> Self {
        match profile {
            BuiltinProfile::PROFILE_X8 => StarkParamsBuilder {
                params_version: 1,
                field: FieldKind::Goldilocks,
                hash: HashKind::Poseidon2 { parameter_set: 0 },
                lde: LdeParams {
                    blowup: 8,
                    order: LdeOrder::RowMajor,
                    coset_tag: 0x5041_5241_4d53_3031,
                },
                fri: FriParams {
                    r: 4,
                    queries: 30,
                    domain_log2: 22,
                    folding: FriFolding::Natural,
                    num_layers: 5,
                },
                merkle: MerkleParams {
                    leaf_encoding: Endianness::Little,
                    leaf_width: 4,
                    arity: MerkleArity::Binary,
                    domain_sep: 0x4d4b_4c5f_5041_5231,
                },
                transcript: TranscriptParams {
                    protocol_tag: 0x5354_524b_5f50_524f,
                    seed: *b"RPP-STARK-PROFILE-X8___________0",
                    challenge_bounds: ChallengeBounds {
                        minimum: 16,
                        maximum: 64,
                    },
                },
                proof: ProofParams {
                    version: 1,
                    max_size_kb: 512,
                },
                security: SecurityBudget {
                    target_bits: 96,
                    soundness_slack_bits: 16,
                },
            },
            BuiltinProfile::PROFILE_HISEC_X16 => StarkParamsBuilder {
                params_version: 2,
                field: FieldKind::Bn254,
                hash: HashKind::Rescue { parameter_set: 1 },
                lde: LdeParams {
                    blowup: 16,
                    order: LdeOrder::ColMajor,
                    coset_tag: 0x5041_5241_4d53_4831,
                },
                fri: FriParams {
                    r: 5,
                    queries: 48,
                    domain_log2: 26,
                    folding: FriFolding::Coset,
                    num_layers: 6,
                },
                merkle: MerkleParams {
                    leaf_encoding: Endianness::Big,
                    leaf_width: 8,
                    arity: MerkleArity::Quaternary,
                    domain_sep: 0x484d_524b_5f48_4953,
                },
                transcript: TranscriptParams {
                    protocol_tag: 0x5354_524b_5f48_4953,
                    seed: *b"RPP-STARK-HISEC-X16___________00",
                    challenge_bounds: ChallengeBounds {
                        minimum: 24,
                        maximum: 96,
                    },
                },
                proof: ProofParams {
                    version: 2,
                    max_size_kb: 768,
                },
                security: SecurityBudget {
                    target_bits: 128,
                    soundness_slack_bits: 32,
                },
            },
        }
    }

    /// Validates the builder fields and emits a [`StarkParams`] instance.
    pub fn build(&self) -> Result<StarkParams, ParamsError> {
        StarkParams::try_from_builder(self)
    }
}

/// Supported built-in profiles.
#[allow(non_camel_case_types)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BuiltinProfile {
    /// Balanced performance profile with blowup 8.
    PROFILE_X8,
    /// High security profile with blowup 16.
    PROFILE_HISEC_X16,
}

impl Default for StarkParamsBuilder {
    fn default() -> Self {
        Self::new()
    }
}
