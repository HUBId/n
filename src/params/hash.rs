use crate::hash::config::BLAKE2S_PARAMETERS_V1;
use crate::hash::deterministic::hash;

use super::ser::serialize_params;
use super::types::HashFamily;
use super::StarkParams;

/// Computes the canonical parameter digest for a [`StarkParams`] instance.
///
/// The function reuses the selected proof hash family for the meta commitment:
/// if the proof profile relies on Poseidon2 or Rescue the same algebraic
/// family is reused with the parameter identifier fixed to `0`.  Otherwise a
/// deterministic Blake2s-style sponge is used.  This keeps the parameter digest
/// aligned with the commitment scheme while remaining deterministic.
pub fn params_hash(params: &StarkParams) -> [u8; 32] {
    let payload = serialize_params(params);
    let mut prefixed = Vec::with_capacity(payload.len() + 8);
    let family = params.hash().family();
    match family {
        HashFamily::Poseidon2 => prefixed.extend_from_slice(b"POSEIDON2/0"),
        HashFamily::Rescue => prefixed.extend_from_slice(b"RESCUE/0"),
        HashFamily::Blake2s => prefixed.extend_from_slice(BLAKE2S_PARAMETERS_V1.domain_tag),
    }
    prefixed.extend_from_slice(&payload);
    hash(&prefixed).into()
}
