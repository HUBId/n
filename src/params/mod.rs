#![forbid(unsafe_code)]

//! Canonical parameter registry for the STARK pipeline.
//!
//! This module defines [`StarkParams`] as the single source of truth for every
//! security and performance relevant parameter in the proving stack.  The
//! structure is intentionally explicit and designed to be serialised into a
//! deterministic byte layout that is shared between the prover, verifier and
//! tooling such as auditors or snapshot tests.
//!
//! # Overview
//!
//! The parameter space is split into themed sub-structures and enums.  The
//! following table summarises the high level groupings:
//!
//! | Group | Description |
//! |-------|-------------|
//! | Field & Hash | Selection of the prime field and hash function family used throughout the pipeline. |
//! | LDE & FRI | Low Degree Extension and Fast Reed–Solomon IOP knobs governing blowup factors and folding strategies. |
//! | Merkle & Transcript | Commitment encoding rules and Fiat–Shamir transcript framing. |
//! | Proof & Security | Envelope size limits and overall soundness budgeting. |
//!
//! All validation logic lives in [`validate`], canonical serialisation lives in
//! [`ser`] and the stable digest computation is exposed via
//! [`StarkParams::params_hash`].  Consumers are expected to use the
//! [`StarkParamsBuilder`] helper which offers safe defaults and pre-defined
//! [`BuiltinProfile`] presets.
//!
//! # Invariants
//!
//! * The canonical serialisation is strictly ordered as documented in
//!   [`ser`] and is stable across Rust versions.
//! * [`StarkParams::params_hash`] commits to this serialisation and therefore
//!   uniquely identifies compatible parameter sets.
//! * [`StarkParams::is_compatible_with`] only allows variations on
//!   non-critical fields, permitting for example different proof size budgets
//!   while still sharing the same security assumptions.
//!
//! # Examples
//!
//! While the module intentionally avoids executable examples to keep the
//! specification deterministic, the tests and snapshot fixtures demonstrate how
//! `BuiltinProfile::PROFILE_X8` and `BuiltinProfile::PROFILE_HISEC_X16` expand
//! into complete parameter sets.

mod builder;
mod hash;
mod ser;
mod stark_params;
mod types;
mod validate;

pub use crate::ser::{SerError, SerKind};
pub use builder::{BuiltinProfile, StarkParamsBuilder};
pub use hash::params_hash;
pub use ser::{deserialize_params, serialize_params};
pub use stark_params::StarkParams;
pub use types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashFamily, HashKind, LdeOrder,
    LdeParams, MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
pub use validate::{ParamsError, ValidationReport};
