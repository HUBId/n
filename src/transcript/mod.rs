//! Fiatâ€“Shamir transcript orchestration for the STARK pipeline.
//!
//! The transcript follows a fixed sequence of phases that bind public inputs,
//! commitments and derived challenges.  Every label is an enum variant to ensure
//! compile-time domain separation.  The canonical order is summarised below:
//!
//! | Phase | Label | Source | Purpose |
//! |-------|-------|--------|---------|
//! | Init | [`TranscriptLabel::ParamsHash`] | [`StarkParams::params_hash`](crate::params::StarkParams::params_hash) | Binds parameter framing. |
//! | Init | [`TranscriptLabel::ProtocolTag`] | `StarkParams::transcript().protocol_tag` | Separates transcript families. |
//! | Init | [`TranscriptLabel::Seed`] | `StarkParams::transcript().seed` | Seeds the deterministic sponge. |
//! | Init | [`TranscriptLabel::ContextTag`] | [`TranscriptContext`] | Domain separation for prover components. |
//! | Public | [`TranscriptLabel::PublicInputsDigest`] | Canonical public input digest | Binds the instance public data. |
//! | TraceCommit | [`TranscriptLabel::TraceRoot`] | Trace commitment root | Pins the execution trace commitment. |
//! | TraceCommit | [`TranscriptLabel::TraceChallengeA`] | Challenge derived from transcript | First algebraic challenge after trace commitment. |
//! | CompCommit | [`TranscriptLabel::CompRoot`] | Constraint commitment root | Binds composition polynomial commitment. |
//! | CompCommit | [`TranscriptLabel::CompChallengeA`] | Transcript challenge | Folding seed for constraint composition. |
//! | FRI | [`TranscriptLabel::FriRoot(i)`](TranscriptLabel::FriRoot) | Layer `i` Merkle root | Commits each FRI layer in sequence. |
//! | FRI | [`TranscriptLabel::FriFoldChallenge(i)`](TranscriptLabel::FriFoldChallenge) | Transcript challenge | Folding randomness for layer `i`. |
//! | Queries | [`TranscriptLabel::QueryCount`] | `StarkParams::fri().queries` | Documents query multiplicity. |
//! | Queries | [`TranscriptLabel::QueryIndexStream`] | Transcript challenges | Index stream for trace/Fri openings. |
//! | Final | [`TranscriptLabel::ProofClose`] | Transcript challenge | Final binding digest stored in the proof. |
//!
//! Determinism guarantee: identical [`StarkParams`] (including `params_hash` and
//! transcript seed), the same [`TranscriptContext`], identical label ordering
//! and payloads yield identical challenge sequences and state digests for both
//! prover and verifier.  Forking via [`Transcript::fork`] produces independent
//! yet deterministic transcripts that inherit the parent state digest.

mod core;
mod ser;
mod types;

pub use core::Transcript;
pub use types::{Felt, TranscriptContext, TranscriptError, TranscriptLabel, TranscriptPhase};
