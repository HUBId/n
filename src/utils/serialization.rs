//! Serialization utilities for proofs and transcripts.
//!
//! The types in this module model the byte-level representation of the
//! proof interface. They are intentionally thin wrappers around byte slices
//! but come with extensive documentation that mirrors the constraints imposed
//! by the proof formats documented elsewhere.

/// Wrapper around proof bytes ensuring explicit conversions.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProofBytes {
    /// Raw proof representation.
    bytes: Vec<u8>,
}

impl ProofBytes {
    /// Creates a new proof byte container.
    pub fn new(bytes: Vec<u8>) -> Self {
        Self { bytes }
    }

    /// Returns the underlying byte slice.
    pub fn as_slice(&self) -> &[u8] {
        &self.bytes
    }
}

/// Opaque wrapper representing a byte encoded witness.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WitnessBlob<'a> {
    /// Borrowed witness bytes provided by the caller.
    pub bytes: &'a [u8],
}

/// Fixed width digest byte array used throughout the documentation layer.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DigestBytes {
    /// Raw digest bytes (e.g. BLAKE3, SHA-256 etc.).
    pub bytes: [u8; 32],
}

/// Wrapper around field element encodings.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FieldElementBytes {
    /// Little endian encoding of a field element.
    pub bytes: [u8; 32],
}
