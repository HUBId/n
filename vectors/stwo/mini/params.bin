0100010003200008000000013130534d415241500440000c000105010102315241505f4c4b4d4f52505f4b5254535250502d535441524b2d50524f46494c452d58385f5f5f5f5f5f5f5f5f5f5f301040010000020000600010
