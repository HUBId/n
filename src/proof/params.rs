use crate::config::{
    ProfileConfig, MERKLE_SCHEME_ID_BLAKE3_4ARY_V1, PROFILE_HIGH_SECURITY_CONFIG,
    PROFILE_STANDARD_CONFIG,
};
use crate::params::{FriFolding, HashKind, LdeOrder, MerkleArity, StarkParams, StarkParamsBuilder};

/// Returns the canonical Stark parameter set shared by prover and verifier.
///
/// The current implementation relies on the deterministic profile used by the
/// built-in prover.  Both parties fix the hashing backend to Blake2s and commit
/// to single-field-element Merkle leaves.  The helper mirrors the prover
/// configuration so callers can rely on identical Merkle framing without
/// leaking the builder details elsewhere in the crate.
pub fn canonical_stark_params(profile: &ProfileConfig) -> StarkParams {
    let mut builder = StarkParamsBuilder::new();
    builder.hash = HashKind::Blake2s { digest_size: 32 };
    builder.merkle.leaf_width = 1;
    if profile.merkle_scheme_id == MERKLE_SCHEME_ID_BLAKE3_4ARY_V1 {
        builder.merkle.arity = MerkleArity::Quaternary;
    } else {
        builder.merkle.arity = MerkleArity::Binary;
    }
    builder.lde.blowup = profile.lde_factor as u32;
    builder.fri.queries = profile.fri_queries;
    builder.fri.domain_log2 = profile.fri_depth_range.max as u16;
    builder.fri.num_layers = profile
        .fri_depth_range
        .max
        .saturating_sub(profile.fri_depth_range.min)
        .saturating_add(1);

    if profile.id == PROFILE_HIGH_SECURITY_CONFIG.id {
        builder.lde.order = LdeOrder::ColMajor;
        builder.fri.folding = FriFolding::Coset;
    } else if profile.id == PROFILE_STANDARD_CONFIG.id {
        builder.lde.order = LdeOrder::RowMajor;
        builder.fri.folding = FriFolding::Natural;
    }

    builder
        .build()
        .expect("canonical Stark parameters must be valid")
}
