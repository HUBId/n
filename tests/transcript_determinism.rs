use insta::assert_json_snapshot;
use proptest::prelude::*;
use rpp_stark::params::{BuiltinProfile, StarkParamsBuilder};
use rpp_stark::transcript::{Transcript, TranscriptContext, TranscriptError, TranscriptLabel};
use rpp_stark::utils::serialization::DigestBytes;

fn sample_params(profile: BuiltinProfile) -> rpp_stark::params::StarkParams {
    StarkParamsBuilder::from_profile(profile)
        .build()
        .expect("profile must be valid")
}

fn sample_digest(byte: u8) -> DigestBytes {
    DigestBytes { bytes: [byte; 32] }
}

#[test]
fn deterministic_state_digest() {
    let params = sample_params(BuiltinProfile::PROFILE_X8);
    let mut t1 = Transcript::new(&params, TranscriptContext::StarkMain);
    let mut t2 = Transcript::new(&params, TranscriptContext::StarkMain);

    let public = sample_digest(1);
    t1.absorb_digest(TranscriptLabel::PublicInputsDigest, &public)
        .unwrap();
    t2.absorb_digest(TranscriptLabel::PublicInputsDigest, &public)
        .unwrap();

    let trace_root = sample_digest(2);
    t1.absorb_digest(TranscriptLabel::TraceRoot, &trace_root)
        .unwrap();
    t2.absorb_digest(TranscriptLabel::TraceRoot, &trace_root)
        .unwrap();

    let trace_challenge_a1 = t1
        .challenge_field(TranscriptLabel::TraceChallengeA)
        .unwrap();
    let trace_challenge_a2 = t2
        .challenge_field(TranscriptLabel::TraceChallengeA)
        .unwrap();
    assert_eq!(trace_challenge_a1, trace_challenge_a2);

    let comp_root = sample_digest(3);
    t1.absorb_digest(TranscriptLabel::CompRoot, &comp_root)
        .unwrap();
    t2.absorb_digest(TranscriptLabel::CompRoot, &comp_root)
        .unwrap();

    let comp_challenge_a1 = t1.challenge_field(TranscriptLabel::CompChallengeA).unwrap();
    let comp_challenge_a2 = t2.challenge_field(TranscriptLabel::CompChallengeA).unwrap();
    assert_eq!(comp_challenge_a1, comp_challenge_a2);

    let mut fri_folds = Vec::new();
    for layer in 0..params.fri().num_layers {
        let fri_root = sample_digest(10 + layer);
        t1.absorb_digest(TranscriptLabel::FriRoot(layer), &fri_root)
            .unwrap();
        t2.absorb_digest(TranscriptLabel::FriRoot(layer), &fri_root)
            .unwrap();
        let fold_a1 = t1
            .challenge_field(TranscriptLabel::FriFoldChallenge(layer))
            .unwrap();
        let fold_a2 = t2
            .challenge_field(TranscriptLabel::FriFoldChallenge(layer))
            .unwrap();
        assert_eq!(fold_a1, fold_a2);
        fri_folds.push(fold_a1);
        // absorb the same digest after comparing to maintain ordering
    }

    let queries = params.fri().queries.to_le_bytes();
    t1.absorb_bytes(TranscriptLabel::QueryCount, &queries)
        .unwrap();
    t2.absorb_bytes(TranscriptLabel::QueryCount, &queries)
        .unwrap();

    let domain_size = 1usize << params.fri().domain_log2;
    let sample_count = 4usize;
    let mut idxs1 = Vec::new();
    let mut idxs2 = Vec::new();
    for _ in 0..sample_count {
        idxs1.push(
            t1.challenge_usize(TranscriptLabel::QueryIndexStream, domain_size)
                .unwrap(),
        );
        idxs2.push(
            t2.challenge_usize(TranscriptLabel::QueryIndexStream, domain_size)
                .unwrap(),
        );
    }
    assert_eq!(idxs1, idxs2);

    let close1 = t1.challenge_bytes(TranscriptLabel::ProofClose, 32).unwrap();
    let close2 = t2.challenge_bytes(TranscriptLabel::ProofClose, 32).unwrap();
    assert_eq!(close1, close2);
    let proof_close_hex: String = close1.iter().map(|b| format!("{:02x}", b)).collect();
    let fri_fold_scalars: Vec<u64> = fri_folds.iter().map(|f| u64::from(*f)).collect();
    assert_json_snapshot!(
        "transcript_profile_x8",
        serde_json::json!({
            "trace_challenge": u64::from(trace_challenge_a1),
            "comp_challenge": u64::from(comp_challenge_a1),
            "fri_folds": fri_fold_scalars,
            "query_indices": idxs1,
            "proof_close_hex": proof_close_hex,
        })
    );
    assert_eq!(t1.state_digest(), t2.state_digest());
}

#[test]
fn different_order_changes_digest() {
    let params = sample_params(BuiltinProfile::PROFILE_X8);
    let mut ordered = Transcript::new(&params, TranscriptContext::StarkMain);
    let mut shuffled = Transcript::new(&params, TranscriptContext::StarkMain);

    let public = sample_digest(4);
    ordered
        .absorb_digest(TranscriptLabel::PublicInputsDigest, &public)
        .unwrap();
    shuffled
        .absorb_digest(TranscriptLabel::PublicInputsDigest, &public)
        .unwrap();

    let trace_root = sample_digest(5);
    ordered
        .absorb_digest(TranscriptLabel::TraceRoot, &trace_root)
        .unwrap();
    shuffled
        .absorb_digest(TranscriptLabel::TraceRoot, &trace_root)
        .unwrap();

    let _ = ordered
        .challenge_field(TranscriptLabel::TraceChallengeA)
        .unwrap();
    let comp_root = sample_digest(6);
    ordered
        .absorb_digest(TranscriptLabel::CompRoot, &comp_root)
        .unwrap();
    let _ = ordered
        .challenge_field(TranscriptLabel::CompChallengeA)
        .unwrap();

    // Shuffled variant skips comp root before taking challenge which must fail.
    let err = shuffled
        .challenge_field(TranscriptLabel::CompChallengeA)
        .unwrap_err();
    assert_eq!(err, TranscriptError::InvalidLabel);
}

proptest! {
    #[test]
    fn usize_challenges_respect_range(range in 1usize..1024usize) {
        let params = sample_params(BuiltinProfile::PROFILE_X8);
        let mut transcript = Transcript::new(&params, TranscriptContext::StarkMain);
        let public = sample_digest(7);
        transcript.absorb_digest(TranscriptLabel::PublicInputsDigest, &public).unwrap();
        let trace = sample_digest(8);
        transcript.absorb_digest(TranscriptLabel::TraceRoot, &trace).unwrap();
        let _ = transcript.challenge_field(TranscriptLabel::TraceChallengeA).unwrap();
        let comp = sample_digest(9);
        transcript.absorb_digest(TranscriptLabel::CompRoot, &comp).unwrap();
        let _ = transcript.challenge_field(TranscriptLabel::CompChallengeA).unwrap();
        for layer in 0..params.fri().num_layers {
            let fri_root = sample_digest(20 + layer);
            transcript
                .absorb_digest(TranscriptLabel::FriRoot(layer), &fri_root)
                .unwrap();
            let _ = transcript
                .challenge_field(TranscriptLabel::FriFoldChallenge(layer))
                .unwrap();
        }
        transcript
            .absorb_bytes(TranscriptLabel::QueryCount, &params.fri().queries.to_le_bytes())
            .unwrap();
        let idx = transcript.challenge_usize(TranscriptLabel::QueryIndexStream, range).unwrap();
        prop_assert!(idx < range);
    }
}
