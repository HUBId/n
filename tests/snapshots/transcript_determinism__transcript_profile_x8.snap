---
source: tests/transcript_determinism.rs
assertion_line: 101
expression: "serde_json::json!({\n    \"trace_challenge\":\n    u64::try_from(trace_challenge_a1).expect(\"canonical trace challenge\"),\n    \"comp_challenge\":\n    u64::try_from(comp_challenge_a1).expect(\"canonical composition challenge\"),\n    \"fri_folds\": fri_fold_scalars, \"query_indices\": idxs1, \"proof_close_hex\":\n    proof_close_hex,\n})"
---
{
  "comp_challenge": 4544395549825592479,
  "fri_folds": [
    7018992374078284144,
    11004696832450032240,
    15614099910591839379,
    9673043499701073733,
    7977881611598001890
  ],
  "proof_close_hex": "2a4b24959fd52a2606420845639724c913cbfdccf9c29045896f590cb9866d5d",
  "query_indices": [
    3789920,
    240205,
    261103,
    2619952
  ],
  "trace_challenge": 18060372420658952913
}
